Main/Source/dungeon.cpp

```diff
@@ -88,47 +91,78 @@ truth dungeon::PrepareLevel(int Index, truth Visual)
+            v2 Displacement = *LevelScript->GetEnterTextDisplacement();
+            game::SetEnterTextDisplacement(Displacement);
+            game::TextScreen(CONST_S("Entering ") + GetLevelDescription(Index)
+                             + CONST_S("...\n\nThis may take some time, please wait."),
+                             Displacement, WHITE, false, 
+                             true, &game::BusyAnimation);
+            game::TextScreen(CONST_S("Entering ") + GetLevelDescription(Index)
+                             + CONST_S("...\n\nPress any key to continue."),
+                             Displacement, WHITE, game::GetAutoPlayMode()<2, 
+                             false, &game::BusyAnimation);
+            game::SetEnterImage(0);
+            delete EnterImage;
+            EnterImage=NULL;
+          }
+          else
+          {
+            game::SetEnterTextDisplacement(ZERO_V2);
+            game::TextScreen(CONST_S("Entering ") + GetLevelDescription(Index)
+                             + CONST_S("...\n\nThis may take some time, please wait."),
+                             ZERO_V2, WHITE, false, true, &game::BusyAnimation);
+          }
+        }
 
-    if(Visual)
-    {
-      if(LevelScript->GetEnterImage())
-      {
-        cbitmap* EnterImage = new bitmap(game::GetDataDir() + "Graphics/" + *LevelScript->GetEnterImage());
-        game::SetEnterImage(EnterImage);
-        v2 Displacement = *LevelScript->GetEnterTextDisplacement();
-        game::SetEnterTextDisplacement(Displacement);
-        game::TextScreen(CONST_S("Entering ") + GetLevelDescription(Index)
-                         + CONST_S("...\n\nThis may take some time, please wait."),
-                         Displacement, WHITE, false, true, &game::BusyAnimation);
-        game::TextScreen(CONST_S("Entering ") + GetLevelDescription(Index)
-                         + CONST_S("...\n\nPress any key to continue."),
-                         Displacement, WHITE, true, false, &game::BusyAnimation);
-        game::SetEnterImage(0);
-        delete EnterImage;
-      }
-      else
-      {
-        game::SetEnterTextDisplacement(ZERO_V2);
-        game::TextScreen(CONST_S("Entering ") + GetLevelDescription(Index)
-                         + CONST_S("...\n\nThis may take some time, please wait."),
-                         ZERO_V2, WHITE, false, true, &game::BusyAnimation);
+        NewLevel->Generate(Index);
+        game::SetCurrentLSquareMap(NewLevel->GetMap());
+        Generated[Index] = true;
+        game::BusyAnimation();
+
+        if(*NewLevel->GetLevelScript()->GenerateMonsters())
+          NewLevel->GenerateNewMonsters(NewLevel->GetIdealPopulation(), false);
+
+        if(genericException::ToggleGenNewLvl())ABORT("expecting gen lvl to become: false");
+        if(fsGenLoopDL!=fsDL)
+          return false; // new level is ok
+      }catch(const genericException& e){
+        // cleanup
+        //TODO it is not working well, memory usage keeps increasing...
```

Comment:
I've been wondering about this strategy and feared that re-generating dungeon level might lead to memory leaks. This might be due [Alloc2D calls in level::Generate](https://github.com/Attnam/ivan/blob/87f4a33fc84e8b3b432c9db4211c3d213433cc98/Main/Source/level.cpp#L327) etc in level.cpp. We will need to have some means of freeing the memory if the level had not been generated successfully, before moving on to trying again.

Reply:
If I remember it well, when a new dungeon level is loaded, the old one is just deleted right? it auto deletes everything inside thru destructors I think. And that doesnt happen so often in a normal gameplay.
But... when put in such loop, for the 250 retry test, after about may be 50 retries you start seeing the difference in memory usage MB by the game.

So, just to fix a normal gameplay, the benefit is better then the memory leak, but that is indeed still a problem.
Alloc2D() is used with 2 variables that I think there could have something missing in some destructor, not necessarily on the below code but may be something deeper:
```
level::~level()
{
  ulong c;

  for(c = 0; c < XSizeTimesYSize; ++c)
    delete NodeMap[0][c];

  for(c = 0; c < Room.size(); ++c)
    delete Room[c];

  delete [] NodeMap;
  delete [] WalkabilityMap;
```

Reply:
I think a profiler can detect and pin point memory leaks in the code right?
The only one I saw running and was very good was from netbeans, but I think it was a java application, despite I think it can profile C++ apps too.

Reply:
> Alloc2D() is used with 2 variables that I think there could have something missing in some destructor, not necessarily on the below code but may be something deeper:

I agree, it probably lies deeper, as it is in all good dungeons...! There's bound to be some extraneous factor leading to leaks elsewhere. 

In benign circumstances, even if the problem is around 1 in 10 games, and the game successfully regenerates a dungeon on the second try, then it is pretty decent. By the time you get to regenerate 50 times, it is probably 1:10^12 games or soemthing like that. 

All the same, it might be good to have a generation counter, as you suggest.

Reply:
Re. the profiler, I think [valgrind](https://www.valgrind.org/) on linux, or similarly the [Tracy profiler](https://github.com/wolfpld/tracy) (cross-platform). The Tracy profiler manual is on the release page: https://github.com/wolfpld/tracy/releases

N.B. Tracy profiler makes use of an immediate-mode gui ([Dear ImGui](https://github.com/ocornut/imgui) by @ocornut) to present results. This makes it very responsive and easy to read.

Reply:
cool, I will try Tracy,

btw, I tested with valgrind 
`valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose --log-file=valgrind-out.txt ./ivan`
the game in wizard mode,
I teleported "|" to some levels to let them be deleted, about 4 times.
but on the log, there is no destructors... no "::~" to be found... so I cant relate memoryleak to a dungeon level deletion from this log...

this was the log summary for that short run:
```
==31254== LEAK SUMMARY:
==31254==    definitely lost: 5,505 bytes in 165 blocks
==31254==    indirectly lost: 13,132,236 bytes in 55,934 blocks
==31254==      possibly lost: 5,103,400 bytes in 7,672 blocks
==31254==    still reachable: 16,509,246 bytes in 7,504 blocks
==31254==                       of which reachable via heuristic:
==31254==                         length64           : 96 bytes in 6 blocks
==31254==                         newarray           : 162,144 bytes in 1,143 blocks
==31254==         suppressed: 0 bytes in 0 blocks
==31254== 
==31254== ERROR SUMMARY: 10000230 errors from 266 contexts (suppressed: 6 from 1)
```
On the full log, there seems to have memory leaks everywhere!
And I think I may have coded one of the biggest ones (at least on a short run) at "main.pp SkipGameScript()" when it analyses the savegame files to gather info to show on the continue game list. I thought stuff I collected were being auto deleted on their destructors, may be I was wrong? may be they just point to data/memory instead of holding the data?

Reply:
mmm... I am having some trouble to compile Tracy, I think it is dependencies, will try again later.